pkg/quotaplugins/quota-forest/quota-manager/quota/quotamanager.go

```diff
@@ -373,7 +377,24 @@ func (m *Manager) AllocateForest(forestName string, consumerID string) (response
 		return nil, fmt.Errorf("consumer %s does not exist, create and add first", consumerID)
 	}
 	if forestController.IsConsumerAllocated(consumerID) {
-		return nil, fmt.Errorf("consumer %s already allocated on forest %s", consumerID, forestName)
+		//allocation can pass if consumers are not released
+		//make sure that appwrapper (consumer) does not exist in target namespace
+		re := regexp.MustCompile("_")
+		split := re.Split(consumerID, -1)
+		namespace := split[1]
+		awName := split[3]
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			klog.Errorf("unable to retrieve in-cluster kubeconfig %v", err)
+		}
+		arbclients := clientset.NewForConfigOrDie(config)
+		_, err = arbclients.ArbV1().AppWrappers(namespace).Get(awName, v1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("consumer %s already allocated on forest %s", consumerID, forestName)
+		} else {
+			//release quota
+			m.DeAllocateForest(forestName, consumerID)
+		}
 	}
```

Comment:
I'm confused here. If the `consumerID` is allocated it is removed in the `else`, but if it's allocated and no AppWrapper is found for the `consumerID` then it isn't deallocated?

Reply:
I'm confused too. The consumer shouldn't exist to begin with. So if it does, then that is an error. I get that from line 393.  But then if it doesn't exist in the forest, then it is deallocated from the forest, where it doesn't exist?